File: nodejs/src/Tree/LCABST/LCABST.ts
import {BaseTreeAlgorithm} from "../BaseTreeAlgorithm";
import {TreeNode} from "../TreeNode";
import {Tree} from "../Tree";

class LCABST extends BaseTreeAlgorithm{
  p: TreeNode | null
  q: TreeNode | null

  constructor(tree: Tree, p: TreeNode | null, q: TreeNode | null) {
    super(tree);
    this.p = p
    this.q = q;
  }

  exec(): TreeNode | null {
    let cur: TreeNode | null = this.tree.root;

    // eslint-disable-next-line no-constant-condition
    while(true){
      // cur node is lees than both
      // @ts-ignore
      if(cur.val < this.p.val && cur.val < this.q.val){
        // @ts-ignore
        cur = cur.right
        // @ts-ignore
      } else if (cur.val > this.p.val && cur.val > this.q.val){
        // @ts-ignore
        cur = cur.left
      } else {
        return cur
      }
    }
  }
}

export default LCABST;

File: go/src/tree/lca_bst/lca_bst.go
package lca_bst

import "algorithms/go/tree"

func lowestCommonAncestor(root, p, q *tree.Node) *tree.Node {
	if root == nil {
		return nil
	}

	for {
		if root.Val < p.Val && root.Val < q.Val {
			root = root.Right
		} else if root.Val > p.Val && root.Val > q.Val {
			root = root.Left
		} else {
			return root
		}
	}

}
